File: backend/src/main/java/com/gxs/bank/controller/LoanController.java
package com.gxs.bank.controller;

import com.gxs.bank.dto.request.LoanApplyRequest;
import com.gxs.bank.dto.request.LoanRepayRequest;
import com.gxs.bank.dto.response.ApiResponse;
import com.gxs.bank.model.Loan;
import com.gxs.bank.model.User;
import com.gxs.bank.service.LoanService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/loans")
@RequiredArgsConstructor
public class LoanController {

    private final LoanService loanService;

    @PostMapping("/apply")
    public ResponseEntity<ApiResponse<Loan>> apply(
            @AuthenticationPrincipal User user,
            @Valid @RequestBody LoanApplyRequest request) {
        Loan loan = loanService.applyForLoan(user.getId(), request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Loan approved", loan));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<Loan>>> getLoans(@AuthenticationPrincipal User user) {
        List<Loan> loans = loanService.getUserLoans(user.getId());
        return ResponseEntity.ok(ApiResponse.ok(loans));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Loan>> getLoan(
            @PathVariable UUID id, @AuthenticationPrincipal User user) {
        Loan loan = loanService.getLoan(id, user.getId());
        return ResponseEntity.ok(ApiResponse.ok(loan));
    }

    @PostMapping("/{id}/repay")
    public ResponseEntity<ApiResponse<Loan>> repay(
            @PathVariable UUID id,
            @AuthenticationPrincipal User user,
            @Valid @RequestBody LoanRepayRequest request) {
        Loan loan = loanService.repay(id, user.getId(), request);
        return ResponseEntity.ok(ApiResponse.ok("Repayment processed", loan));
    }

    @GetMapping("/calculate")
    public ResponseEntity<ApiResponse<Map<String, Object>>> calculate(
            @RequestParam BigDecimal amount,
            @RequestParam(defaultValue = "1.08") BigDecimal rate,
            @RequestParam int tenureMonths) {
        Map<String, Object> result = loanService.calculateLoan(amount, rate, tenureMonths);
        return ResponseEntity.ok(ApiResponse.ok(result));
    }
}

File: backend-python/com/gxs/bank/controller/LoanController.py
from decimal import Decimal

from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse

from com.gxs.bank.config.database import get_db
from com.gxs.bank.config.serialization import serialize
from com.gxs.bank.dto.request.LoanApplyRequest import LoanApplyRequest
from com.gxs.bank.dto.request.LoanRepayRequest import LoanRepayRequest
from com.gxs.bank.dto.response.ApiResponse import ok
from com.gxs.bank.security.JwtAuthenticationFilter import get_current_user
from com.gxs.bank.service.LoanService import LoanService


router = APIRouter(prefix="/api/loans", tags=["Loan"])


@router.post("/apply")
def applyLoan(request: LoanApplyRequest, current_user=Depends(get_current_user), db=Depends(get_db)):
    loan = LoanService(db).applyForLoan(current_user.id, request)
    return JSONResponse(status_code=201, content=ok(serialize(loan), "Loan approved"))


@router.get("")
def getLoans(current_user=Depends(get_current_user), db=Depends(get_db)):
    loans = LoanService(db).getUserLoans(current_user.id)
    return ok(serialize(loans))


@router.get("/{loan_id}")
def getLoan(loan_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    loan = LoanService(db).getLoan(loan_id, current_user.id)
    return ok(serialize(loan))


@router.post("/{loan_id}/repay")
def repayLoan(loan_id: str, request: LoanRepayRequest, current_user=Depends(get_current_user), db=Depends(get_db)):
    loan = LoanService(db).repay(loan_id, current_user.id, request)
    return ok(serialize(loan), "Repayment processed")


@router.get("/calculate")
def calculateLoan(
    amount: Decimal = Query(...),
    rate: Decimal = Query(default=Decimal("1.08")),
    tenureMonths: int = Query(...),
    db=Depends(get_db),
):
    result = LoanService(db).calculateLoan(amount, rate, tenureMonths)
    return ok(serialize(result))
